### PR TITLE
layers: Add 03675 03676

### PR DESCRIPTION
**Pull request by arno-lunarg**
Part of https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/3792

**VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03675**
For each pInfos[i], dstAccelerationStructure must have been created with a value of [VkAccelerationStructureCreateInfoKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkAccelerationStructureCreateInfoKHR.html)::size greater than or equal to the memory size required by the build operation, as returned by [vkGetAccelerationStructureBuildSizesKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkGetAccelerationStructureBuildSizesKHR.html) with pBuildInfo = pInfos[i] and with each element of the pMaxPrimitiveCounts array greater than or equal to the equivalent ppBuildRangeInfos[i][j].primitiveCount values for j in [0,pInfos[i].geometryCount)

**VUID-vkCmdBuildAccelerationStructuresKHR-ppBuildRangeInfos-03676**
Each element of ppBuildRangeInfos[i] must be a valid pointer to an array of pInfos[i].geometryCount VkAccelerationStructureBuildRangeInfoKHR structures

### CODE REVIEW

File: layers/core_checks/cc_ray_tracing.cpp [hunk 1]
```diff
@@ -491,8 +491,17 @@ bool CoreChecks::ValidateAccelerationStructuresMemoryAlisasing(VkCommandBuffer c
 
 bool CoreChecks::ValidateAccelerationBuffers(uint32_t info_i, const VkAccelerationStructureBuildGeometryInfoKHR &info,
                                              const VkAccelerationStructureBuildRangeInfoKHR *geometry_build_ranges,
-                                             const Location &loc) const {
+                                             const Location &info_loc) const {
     bool skip = false;
+
+    if (!geometry_build_ranges) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_ray_tracing.cpp:497]: this should be done in `StatelessValidation::manual_PreCallValidateCmdBuildAccelerationStructuresKHR` using `ValidateArray`

(same as `VUID-vkCmdBuildAccelerationStructuresKHR-ppBuildRangeInfos-parameter` but without the count VU)

```c++
// something like
skip |= ValidateArray(loc, loc.dot(Field::ppBuildRangeInfos, i), pInfos[i].geometryCount, &ppBuildRangeInfos[i], false, true, kVUIDUndefined, "VUID-vkCmdBuildAccelerationStructuresKHR-ppBuildRangeInfos-03676");
```

File: layers/core_checks/cc_ray_tracing.cpp [hunk 2]
```diff
@@ -578,10 +587,22 @@ bool CoreChecks::ValidateAccelerationBuffers(uint32_t info_i, const VkAccelerati
         }
     }
 
+    if (const auto dst_as_state = Get<vvl::AccelerationStructureKHR>(info.dstAccelerationStructure)) {
+        const VkDeviceSize as_minimum_size = rt::ComputeAccelerationStructureSize(device, info, geometry_build_ranges);
+        if (dst_as_state->create_infoKHR.size < as_minimum_size) {
+            const LogObjectList objlist(info.dstAccelerationStructure);
+            skip |= LogError("VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03675", objlist,
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_ray_tracing.cpp:594]:
```suggestion
            skip |= LogError("VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03675", info.dstAccelerationStructure,
```

File: layers/core_checks/cc_ray_tracing.cpp (hunk 1 above)
Comment (arno-lunarg) [reply layers/core_checks/cc_ray_tracing.cpp:497]: Ok!

File: layers/core_checks/cc_ray_tracing.cpp (hunk 2 above)
Comment (arno-lunarg) [reply layers/core_checks/cc_ray_tracing.cpp:594]: ah you make me realise I had forgotten to add the command buffer to the object list

File: layers/core_checks/cc_ray_tracing.cpp [hunk 3]
```diff
@@ -578,10 +580,22 @@ bool CoreChecks::ValidateAccelerationBuffers(uint32_t info_i, const VkAccelerati
         }
     }
 
+    if (const auto dst_as_state = Get<vvl::AccelerationStructureKHR>(info.dstAccelerationStructure)) {
+        const VkDeviceSize as_minimum_size = rt::ComputeAccelerationStructureSize(device, info, geometry_build_ranges);
+        if (dst_as_state->create_infoKHR.size < as_minimum_size) {
+            const LogObjectList objlist(cmd_buffer, info.dstAccelerationStructure);
+            skip |= LogError("VUID-vkCmdBuildAccelerationStructuresKHR-pInfos-03675", objlist,
+                             info_loc.dot(Field::dstAccelerationStructure),
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_ray_tracing.cpp:588]: this english here seems a bit hard to follow

did you mean

> was created with size (..), but built with corresponding ppBuildRangeInfos[..] that requires a minimum size of (..)

File: layers/core_checks/cc_ray_tracing.cpp [hunk 4]
```diff
@@ -489,10 +489,12 @@ bool CoreChecks::ValidateAccelerationStructuresMemoryAlisasing(VkCommandBuffer c
     return skip;
 }
 
-bool CoreChecks::ValidateAccelerationBuffers(uint32_t info_i, const VkAccelerationStructureBuildGeometryInfoKHR &info,
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_ray_tracing.cpp:492]: while adding the `VkCommandBuffer` here , can we just make every error in `ValidateAccelerationBuffers` have this in its object list

File: layers/stateless/sl_ray_tracing.cpp [hunk 5]
```diff
@@ -1320,7 +1320,14 @@ bool StatelessValidation::manual_PreCallValidateCmdBuildAccelerationStructuresKH
                 }
             }
         }
+
+        skip |= ValidateArray(info_loc.dot(Field::geometryCount), error_obj.location.dot(Field::ppBuildRangeInfos, info_i),
+                              info->geometryCount, &ppBuildRangeInfos[info_i], false, true, kVUIDUndefined,
+                              "VUID-vkCmdBuildAccelerationStructuresKHR-ppBuildRangeInfos-03676");
     }
+
+    // something like
```
Comment (spencer-lunarg) [inline layers/stateless/sl_ray_tracing.cpp:1329]: ?

Comment (arno-lunarg) [reply layers/stateless/sl_ray_tracing.cpp:1329]: ¯\(°_o)/¯

File: layers/core_checks/cc_ray_tracing.cpp (hunk 3 above)
Comment (arno-lunarg) [reply layers/core_checks/cc_ray_tracing.cpp:616]: I updated the error message slightly, tell me what you think!

File: tests/framework/ray_tracing_objects.cpp [hunk 6]
```diff
@@ -13,6 +13,10 @@
 
 #include "utils/vk_layer_utils.h"
 
+#include <iostream>
+
+#define DEBUG_LOG_GET_SIZE_INFO 0
```
Comment (spencer-lunarg) [inline tests/framework/ray_tracing_objects.cpp:18]: I rather not just be adding various new debug `printf` style logging here, can we pull this out to another PR... agree on the idea, have opinions on the implementation but rather not have that bog down getting these VUs in  first

Comment (arno-lunarg) [reply tests/framework/ray_tracing_objects.cpp:18]: ok!

**Review by spencer-lunarg — APPROVED** [review]
